cpp/ql/lib/semmle/code/cpp/security/InvalidPointerDereference/InvalidPointerToDereference.qll

```diff
@@ -2,6 +2,73 @@
  * This file provides the second phase of the `cpp/invalid-pointer-deref` query that identifies flow
  * from the out-of-bounds pointer identified by the `AllocationToInvalidPointer.qll` library to
  * a dereference of the out-of-bounds pointer.
+ *
+ * Consider the following snippet:
+ * ```cpp
+ * 1. char* base = (char*)malloc(size);
+ * 2. char* end = base + size;
+ * 3. for(char *p = base; p <= end; p++) {
+ * 4.   use(*p); // BUG: Should have been bounded by `p < end`.
+ * 5. }
+ * ```
+ * this file identifies the flow from `base + size` to `end`. We call `base + size` the "dereference source" and `end`
+ * the "dereference sink" (even though `end` is not actually dereferenced we will use this term because we will perform
+ * dataflow to find a use of a pointer `x` such that `x <= end` which is dereferenced. In the above example, `x` is `p`
+ * on line 4).
+ *
+ * Merely _constructing_ a pointer that's out-of-bounds is fine if the pointer is never dereferenced (in reality, the
+ * standard only guarentees that it's safe to move the pointer one element past the last element. But we ignore that
+ * here). So this step is about identifying which of those out-of-bounds pointers found by `pointerAddInstructionHasBounds`
+ * in `AllocationToInvalidPointer.qll` that are actually being dereferenced. We do this using a regular dataflow
+ * configuration (see `InvalidPointerToDerefConfig`).
+ *
+ * This dataflow traversal defines the set of sources as any dataflow node `n` such that there exists a pointer-arithmetic
+ * instruction `pai` found by `AllocationToInvalidPointer.qll` and `n.asInstruction() >= pai + deltaDerefSourceAndPai`.
+ * Here, `deltaDerefSourceAndPai` is the constant difference between the source we track for finding a dereference and the
+ * pointer-arithmetic instruction.
+ *
+ * The set of sinks is defined as any dataflow node `n` such that `addr <= n.asInstruction() + deltaDerefSinkAndDerefAddress`
+ * for some address operand `addr` and constant difference `deltaDerefSinkAndDerefAddress`. Since an address operand is
+ * always consumed by an instruction that performs a dereference this lets us identify a "bad dereference". We call the
+ * instruction that consumes the address operand the "operation".
+ *
+ * For example, consider the flow from `base + size` to `end` above. The sink is `end` on line 3 because that is a dataflow
+ * node whose underlying instruction non-strictly upper bounds the address operand `p` in `use(*p)`. The load attached to `*p`
+ * is the "operation". To ensure that the path makes intuitive sense, we only pick operations that are control-flow reachable
+ * from the dereference sink.
+ *
+ * To compute the amount of the dereference is away from the final entry of the allocation, we sum the two deltas `delta1` and
+ * `delta2`. This is done in the `operationIsOffBy` predicate (which is the only predicate exposed by this file).
```

Comment:
I think `delta1` and `delta` need to be renamed here? Is it actually correct that we're summing two things here and not three (the delta at the sink of the product flow (be amount `r` is bigger than `b`), the delta at the source here, and the delta at the sink here).

Reply:
I've fixed up the names in 46832d0b170bc8aacee6882894ccfc05c883c1ca. We're indeed only summing those two latter numbers in the `operationIsOffBy` predicate, and the first constant (i.e., the one from product flow) is existentially quantified in `invalidPointerToDerefSource` 🤔. I'll note this down in an issue and see if I can come up with a case where this gives an incorrect alert.